FreeBSD pthread_getaffinity_np not detected

On FreeBSD 13.2, 64 bit x86 or 64 bit ARM, s.h does not define `HAS_BSD_GETAFFINITY_NP` as expected.  The function `pthread_getaffinity_np` is not found because command line to compile the test program needs -pthread or -lpthreads.

```
configure:18103: cc -o conftest    conftest.c -lm  >&5
ld: error: undefined symbol: pthread_getaffinity_np
>>> referenced by conftest.c
>>>               /tmp/conftest-2a7a19.o:(main)
cc: error: linker command failed with exit code 1 (use -v to see invocation)
```

With the new branch `pthread_getaffinity_np` is detected but arguably it works for the wrong reason.

On FreeBSD 12, which is nearing end of life, the patch works as intended.  The GNU version does not compile and the BSD version does.  `HAS_BSD_GETAFFINITY_NP` is defined.

On FreeBSD 13 and 14 configure incorrectly detects the GNU flavor of `pthread_getaffinity_np`.  The FreeBSD implementation has become sufficiently compatible that the runtime compiles without warning.   If there is a test that depends on `pthread_getaffinity_np` working in the runtime I can run it.

A missing function declaration is a warning rather than an error.
```
configure:19281: checking arguments for pthread_getaffinity_np
configure:19299: cc -o conftest  -pthread   conftest.c -lm  -lpthread >&5
conftest.c:145:7: warning: call to undeclared function 'pthread_getaffinity_np'; ISO C99
and later do not support implicit function declarations [-Wimplicit-function-declaration]
      pthread_getaffinity_np(pthread_self(), sizeof(cs), &cs);
      ^
1 warning generated.
configure:19299: $? = 0
configure:19301: result: gnu
```
This warning does not appear compiling the runtime because domain.c includes the right set of header files.

I don't have any other BSD flavors to test on.

Many thanks for for your detailed feedback.

John F. Carr (2023/07/27 13:11 -0700):
> On FreeBSD 13 and 14 configure incorrectly detects the GNU flavor of
> `pthread_getaffinity_np`.  The FreeBSD implementation has become
> sufficiently compatible that the runtime compiles without warning.
> If there is a test that depends on `pthread_getaffinity_np` working in
> the runtime I can run it.

This function is called by the runtime primitive
`caml_recommended_domain_count`, which is itself called by the
`Domain.Raw.get_recommended_domain_count` function from the standard
library, in turn called by `Domain.recommended_domain_count`.

I think the test in
`testsuite/tests/parallel/recommended_domain_count.ml` exercises this.
You can run it directly e.g. by doing, from the toplevel of the source tree:

./ocamltest/ocamltest.opt testsuite/tests/parallel/recommended_domain_count.ml

> A missing function declaration is a warning rather than an error.
> ```
> configure:19281: checking arguments for pthread_getaffinity_np
> configure:19299: cc -o conftest  -pthread   conftest.c -lm  -lpthread >&5
> conftest.c:145:7: warning: call to undeclared function 'pthread_getaffinity_np'; ISO C99
> and later do not support implicit function declarations [-Wimplicit-function-declaration]
>       pthread_getaffinity_np(pthread_self(), sizeof(cs), &cs);
>       ^
> 1 warning generated.
> configure:19299: $? = 0
> configure:19301: result: gnu
> ```
> This warning does not appear compiling the runtime because domain.c
> includes the right set of header files.

OK, is that to say that the configure test is also missing the right set
of headers and that the warning we currently see can not be used because
it's there just because we fail to include the right headers?

I'm wondering, thus, how the BSD implementation is supossed to be
distinguished for the GNU one or even whether it should be
distinguished, if efforts have been made to makeit compatible enough,
perhaps we can just make this work? Perhaps it does actually already
work and the only fix is to add the right headers to the configure test?

Let's see whether the test mentionned above passes?

--- a/configure
+++ b/configure
@@ -18260,65 +18260,6 @@ then :
 fi
 
 
-## pthread_getaffinity_np, args differ from GNU and BSD
-{ printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking arguments for pthread_getaffinity_np" >&5
-printf %s "checking arguments for pthread_getaffinity_np... " >&6; }
-cat confdefs.h - <<_ACEOF >conftest.$ac_ext
-/* end confdefs.h.  */
-#define _GNU_SOURCE
-      #include <sched.h>
-      #include <pthread.h>
-int
-main (void)
-{
-cpu_set_t cs;
-      CPU_ZERO(&cs);
-      CPU_COUNT(&cs);
-      pthread_getaffinity_np(pthread_self(), sizeof(cs), &cs);
-  ;
-  return 0;
-}
-_ACEOF
-if ac_fn_c_try_link "$LINENO"
-then :
-  { printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: gnu" >&5
-printf "%s\n" "gnu" >&6; }
-  printf "%s\n" "#define HAS_GNU_GETAFFINITY_NP 1" >>confdefs.h
-
-else $as_nop
-  cat confdefs.h - <<_ACEOF >conftest.$ac_ext
-/* end confdefs.h.  */
-#include <pthread.h>
-        #include <pthread_np.h>
-        #include <sys/cpuset.h>
-int
-main (void)
-{
-cpuset_t cs;
-        /* Not every BSD has CPU_ZERO and CPU_COUNT (NetBSD) */
-        CPU_ZERO(&cs);
-        CPU_COUNT(&cs);
-        pthread_getaffinity_np(pthread_self(), sizeof(cs), &cs);
-  ;
-  return 0;
-}
-_ACEOF
-if ac_fn_c_try_link "$LINENO"
-then :
-  { printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: BSD" >&5
-printf "%s\n" "BSD" >&6; }
-    printf "%s\n" "#define HAS_BSD_GETAFFINITY_NP 1" >>confdefs.h
-
-else $as_nop
-  { printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: pthread_getaffinity_np not found" >&5
-printf "%s\n" "pthread_getaffinity_np not found" >&6; }
-fi
-rm -f core conftest.err conftest.$ac_objext conftest.beam \
-    conftest$ac_exeext conftest.$ac_ext
-fi
-rm -f core conftest.err conftest.$ac_objext conftest.beam \
-    conftest$ac_exeext conftest.$ac_ext
-
 if test -n "$ac_tool_prefix"; then
   # Extract the first word of "${ac_tool_prefix}pkg-config", so it can be a program name with args.
 set dummy ${ac_tool_prefix}pkg-config; ac_word=$2
@@ -19339,6 +19280,65 @@ ac_compiler_gnu=$ac_cv_c_compiler_gnu
 
  ;;
 esac
+
+## pthread_getaffinity_np, args differ from GNU and BSD
+{ printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking arguments for pthread_getaffinity_np" >&5
+printf %s "checking arguments for pthread_getaffinity_np... " >&6; }
+cat confdefs.h - <<_ACEOF >conftest.$ac_ext
+/* end confdefs.h.  */
+#define _GNU_SOURCE
+      #include <sched.h>
+      #include <pthread.h>
+int
+main (void)
+{
+cpu_set_t cs;
+      CPU_ZERO(&cs);
+      CPU_COUNT(&cs);
+      pthread_getaffinity_np(pthread_self(), sizeof(cs), &cs);
+  ;
+  return 0;
+}
+_ACEOF
+if ac_fn_c_try_link "$LINENO"
+then :
+  { printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: GNU" >&5
+printf "%s\n" "GNU" >&6; }
+  printf "%s\n" "#define HAS_GNU_GETAFFINITY_NP 1" >>confdefs.h
+
+else $as_nop
+  cat confdefs.h - <<_ACEOF >conftest.$ac_ext
+/* end confdefs.h.  */
+#include <pthread.h>
+        #include <pthread_np.h>
+        #include <sys/cpuset.h>
+int
+main (void)
+{
+cpuset_t cs;
+        /* Not every BSD has CPU_ZERO and CPU_COUNT (NetBSD) */
+        CPU_ZERO(&cs);
+        CPU_COUNT(&cs);
+        pthread_getaffinity_np(pthread_self(), sizeof(cs), &cs);
+  ;
+  return 0;
+}
+_ACEOF
+if ac_fn_c_try_link "$LINENO"
+then :
+  { printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: BSD" >&5
+printf "%s\n" "BSD" >&6; }
+    printf "%s\n" "#define HAS_BSD_GETAFFINITY_NP 1" >>confdefs.h
+
+else $as_nop
+  { printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: pthread_getaffinity_np not found" >&5
+printf "%s\n" "pthread_getaffinity_np not found" >&6; }
+fi
+rm -f core conftest.err conftest.$ac_objext conftest.beam \
+    conftest$ac_exeext conftest.$ac_ext
+fi
+rm -f core conftest.err conftest.$ac_objext conftest.beam \
+    conftest$ac_exeext conftest.$ac_ext
 
 ## Activate the systhread library
 

--- a/configure
+++ b/configure
@@ -19287,6 +19287,9 @@ cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 #define _GNU_SOURCE
       #include <sched.h>
+      #ifdef HAS_PTHREAD_NP_H
+      #include <pthread_np.h>
+      #endif
       #include <pthread.h>
 int
 main (void)

--- a/configure
+++ b/configure
@@ -19254,8 +19254,9 @@ test -n "$PTHREAD_CXX" || PTHREAD_CXX="$CXX"
 # Finally, execute ACTION-IF-FOUND/ACTION-IF-NOT-FOUND:
 if test "x$ax_pthread_ok" = "xyes"; then
         common_cflags="$common_cflags $PTHREAD_CFLAGS"
-    saved_CFLAGS="$CFLAGS"
-    saved_LIBS="$LIBS"
+    # The two following lines add flags and libraries for pthread to the
+    # global CFLAGS and LIBS variables. This means that all the subsequent
+    # tests can rely on the assumption that pthread is enabled.
     CFLAGS="$CFLAGS $PTHREAD_CFLAGS"
     LIBS="$LIBS $PTHREAD_LIBS"
     ac_fn_c_check_func "$LINENO" "sigwait" "ac_cv_func_sigwait"
@@ -19265,8 +19266,6 @@ then :
 
 fi
 
-    LIBS="$saved_LIBS"
-    CFLAGS="$saved_CFLAGS"
         :
 else
         ax_pthread_ok=no

--- a/configure.ac
+++ b/configure.ac
@@ -2161,6 +2161,9 @@ AC_LINK_IFELSE(
   [AC_LANG_PROGRAM(
     [[#define _GNU_SOURCE
       #include <sched.h>
+      #ifdef HAS_PTHREAD_NP_H
+      #include <pthread_np.h>
+      #endif
       #include <pthread.h>]],
     [[cpu_set_t cs;
       CPU_ZERO(&cs);

--- a/configure.ac
+++ b/configure.ac
@@ -2078,33 +2078,6 @@ AC_CHECK_HEADER([spawn.h],
 AC_CHECK_FUNC([ffs], [AC_DEFINE([HAS_FFS])])
 AC_CHECK_FUNC([_BitScanForward], [AC_DEFINE([HAS_BITSCANFORWARD])])
 
-## pthread_getaffinity_np, args differ from GNU and BSD
-AC_MSG_CHECKING([arguments for pthread_getaffinity_np])
-AC_LINK_IFELSE(
-  [AC_LANG_PROGRAM(
-    [[#define _GNU_SOURCE
-      #include <sched.h>
-      #include <pthread.h>]],
-    [[cpu_set_t cs;
-      CPU_ZERO(&cs);
-      CPU_COUNT(&cs);
-      pthread_getaffinity_np(pthread_self(), sizeof(cs), &cs);]])],
-  [AC_MSG_RESULT([gnu])
-  AC_DEFINE([HAS_GNU_GETAFFINITY_NP])],
-  [AC_LINK_IFELSE(
-    [AC_LANG_PROGRAM(
-      [[#include <pthread.h>
-        #include <pthread_np.h>
-        #include <sys/cpuset.h>]],
-      [[cpuset_t cs;
-        /* Not every BSD has CPU_ZERO and CPU_COUNT (NetBSD) */
-        CPU_ZERO(&cs);
-        CPU_COUNT(&cs);
-        pthread_getaffinity_np(pthread_self(), sizeof(cs), &cs);]])],
-    [AC_MSG_RESULT([BSD])
-    AC_DEFINE([HAS_BSD_GETAFFINITY_NP])],
-    [AC_MSG_RESULT([pthread_getaffinity_np not found])])])
-
 AC_PATH_TOOL([PKG_CONFIG], [pkg-config], [false])
 
 ## ZSTD compression library
@@ -2182,6 +2155,33 @@ AS_CASE([$host],
     CFLAGS="$saved_CFLAGS"],
     [AC_MSG_ERROR(m4_normalize([POSIX threads are required but not supported on
       this platform]))])])
+
+## pthread_getaffinity_np, args differ from GNU and BSD
+AC_MSG_CHECKING([arguments for pthread_getaffinity_np])
+AC_LINK_IFELSE(
+  [AC_LANG_PROGRAM(
+    [[#define _GNU_SOURCE
+      #include <sched.h>
+      #include <pthread.h>]],
+    [[cpu_set_t cs;
+      CPU_ZERO(&cs);
+      CPU_COUNT(&cs);
+      pthread_getaffinity_np(pthread_self(), sizeof(cs), &cs);]])],
+  [AC_MSG_RESULT([GNU])
+  AC_DEFINE([HAS_GNU_GETAFFINITY_NP])],
+  [AC_LINK_IFELSE(
+    [AC_LANG_PROGRAM(
+      [[#include <pthread.h>
+        #include <pthread_np.h>
+        #include <sys/cpuset.h>]],
+      [[cpuset_t cs;
+        /* Not every BSD has CPU_ZERO and CPU_COUNT (NetBSD) */
+        CPU_ZERO(&cs);
+        CPU_COUNT(&cs);
+        pthread_getaffinity_np(pthread_self(), sizeof(cs), &cs);]])],
+    [AC_MSG_RESULT([BSD])
+    AC_DEFINE([HAS_BSD_GETAFFINITY_NP])],
+    [AC_MSG_RESULT([pthread_getaffinity_np not found])])])
 
 ## Activate the systhread library
 

--- a/configure.ac
+++ b/configure.ac
@@ -2146,13 +2146,12 @@ AS_CASE([$host],
     [PTHREAD_LIBS="-l:libpthread.lib"],
   [AX_PTHREAD(
     [common_cflags="$common_cflags $PTHREAD_CFLAGS"
-    saved_CFLAGS="$CFLAGS"
-    saved_LIBS="$LIBS"
+    # The two following lines add flags and libraries for pthread to the
+    # global CFLAGS and LIBS variables. This means that all the subsequent
+    # tests can rely on the assumption that pthread is enabled.
     CFLAGS="$CFLAGS $PTHREAD_CFLAGS"
     LIBS="$LIBS $PTHREAD_LIBS"
-    AC_CHECK_FUNC([sigwait], [AC_DEFINE([HAS_SIGWAIT])])
-    LIBS="$saved_LIBS"
-    CFLAGS="$saved_CFLAGS"],
+    AC_CHECK_FUNC([sigwait], [AC_DEFINE([HAS_SIGWAIT])])],
     [AC_MSG_ERROR(m4_normalize([POSIX threads are required but not supported on
       this platform]))])])
 

